src/test/java/org/jenkinsci/plugins/workflow/ArtifactManagerTest.java

```diff
@@ -0,0 +1,395 @@
+    }
+    private static class FindEncoding extends MasterToSlaveCallable<String, Exception> {
+        @Override public String call() throws Exception {
+            return System.getProperty("file.encoding") + " vs. " + System.getProperty("sun.jnu.encoding");
+        }
+    }
+
+    @FunctionalInterface
+    interface TestFunction {
+        void apply(DumbSlave agent, FreeStyleProject p, FreeStyleBuild b, FilePath ws) throws Exception;
+    }
+    @FunctionalInterface
+    interface TestStashFunction {
+        void apply(FreeStyleProject p, FreeStyleBuild b, FilePath ws, Launcher launcher, EnvVars env,
+                TaskListener listener) throws Exception;
+    }
+
+    private static class StashFunction implements TestFunction {
+        private JenkinsRule r;
+        private boolean weirdCharacters;
+        private TestStashFunction f;
+
+        StashFunction(@Nonnull JenkinsRule r, boolean weirdCharacters, TestStashFunction f) {
+            this.r = r;
+            this.weirdCharacters = weirdCharacters;
+            this.f = f;
+        }
+
+        @Override
+        public void apply(DumbSlave agent, FreeStyleProject p, FreeStyleBuild b, FilePath ws) throws Exception {
+            TaskListener listener = StreamTaskListener.fromStderr();
+            Launcher launcher = agent.createLauncher(listener);
+            EnvVars env = agent.toComputer().getEnvironment();
+            env.putAll(agent.toComputer().buildEnvironment(listener));
+            // Make sure we can stash and then unstash within a build:
+            StashManager.stash(b, "stuff", ws, launcher, env, listener, "file", null, false, false);
+            ws.child("file").delete();
+            StashManager.unstash(b, "stuff", ws, launcher, env, listener);
+            assertEquals("content", ws.child("file").readToString());
+            ws.child("file").delete();
+            // Copy stashes and artifacts from one build to a second one:
+            p.getPublishersList().clear();
+            FreeStyleBuild b2 = r.buildAndAssertSuccess(p);
+            ExtensionList.lookupSingleton(StashManager.CopyStashesAndArtifacts.class).copy(b, b2, listener);
+            // Verify the copied stashes:
+            StashManager.unstash(b2, "stuff", ws, launcher, env, listener);
+            assertEquals("content", ws.child("file").readToString());
+            // And the copied artifacts:
+            VirtualFile root = b2.getArtifactManager().root();
+            new Verify(agent, root, weirdCharacters).run();
+            // Also delete the original:
+            StashManager.clearAll(b, listener);
+            // Stashes should have been deleted, but not artifacts:
+            assertTrue(b.getArtifactManager().root().child("file").isFile());
+            ws.deleteContents();
+            assertFalse(ws.child("file").exists());
+            f.apply(p, b, ws, launcher, env, listener);
+        }
+    }
+
+    private static class Verify {
```

Comment:
This usage feels... very nonintuitive and hard to follow.  Please, could we do something to clean it up?

Reply:
It is somewhat hard to follow, but I do not see any particular simplification.